---
title: "bulk-rnaseq_setbp1"
author: "Gokberk Alagoz, Maggie Wong, L&G Dept., MPI for Psycholinguistics"
date: "`r Sys.Date()`"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
# all in R/4.0.3 using vnc server
knitr::opts_chunk$set(echo = TRUE)
library(Rsubread)
library(biomaRt)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(DESeq2)
library(vsn)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(ComplexHeatmap)
library(circlize)
library(cowplot)
library(topGO)
library(VennDiagram)
library("ggbeeswarm")
library(EnsDb.Hsapiens.v79)
library(RRHO)
```

## 

```{r data_prep}
#deseq_outDir = "/data/workspaces/lag/workspaces/lg-scrna/working_data/gokberk/setbp1/results/DESeq/"

if (Sys.info()['sysname']=='Windows') {dir="P://workspaces/"} else {dir="/data/workspaces/lag/workspaces/"}
deseq_outDir = paste(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/diffexpression/MW_20250528/results/DESeq/",sep="")

#always save RData
#save.image(file = paste0(dir, "lg-scrna/working_data/Maggie/working/script/bulk_rna_SETBP1_HDF/from_gokberk/20250528/setbp1_hdf_rna_20250528_final.RData"))

#load(file = paste0(dir, "lg-scrna/working_data/Maggie/working/script/bulk_rna_SETBP1_HDF/from_gokberk/20250528/setbp1_hdf_rna_20250528_final.RData"))

# Read bam files
list_of_bam1 = list.files(path = "/data/workspaces/lag/workspaces/lg-scrna/primary_data/maggie_data_release/setbp1_hdf_b1_rna/X201SC19120602-Z01-F003/bam", pattern = ".bam$", full.names = T)
list_of_bam2 = list.files(path = "/data/workspaces/lag/workspaces/lg-scrna/primary_data/maggie_data_release/20211123_setbp1_hdf_b2_rna/X201SC21072043-Z01-F002/bam", pattern = ".bam$", full.names = T)

list_of_bam = c(list_of_bam1,list_of_bam2)

# Create a count table
rsubread.counts = featureCounts(list_of_bam,
                                annot.ext="/data/workspaces/lag/workspaces/lg-scrna/primary_data/maggie_data_release/20211123_setbp1_hdf_b2_rna/X201SC21072043-Z01-F002/ref/genome.gtf/genome.gtf",
                                isGTFAnnotationFile = TRUE,
                                GTF.featureType = "exon",
                                GTF.attrType = "gene_id",
                                useMetaFeatures = TRUE,
                                allowMultiOverlap = FALSE,
                                isPairedEnd = TRUE,
                                requireBothEndsMapped = TRUE)


# Save the count table as an R object
save(rsubread.counts, file = "/data/workspaces/lag/workspaces/lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/featureCounts/rsubread.counts.RData")
load("/data/workspaces/lag/workspaces/lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/featureCounts/rsubread.counts.RData")
load(file = paste0(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/featureCounts/rsubread.counts.RData"))

# Get read counts per gene
rcounts = rsubread.counts$counts
# Get gene annotations
annot = rsubread.counts$annotation

head(rcounts)
dim(rcounts)
# there are 58,735 genes and 69 samples

# Load the metadata of your samples
inf = read.table(file = paste0(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/sample_name_condition_v3-2.csv"), skipNul = T, header = T, sep=",")

```

## 

```{r prep4deseq}

# Check the mean gene length
summary(annot$Length)

# Take out all transcripts that have no reads in any of the experiments
rcounts = rcounts[(rowSums(rcounts)>0),]

# Make countdata a dataframe
countdata = data.frame(GeneID = rownames(rcounts), rcounts)

# Add geneLength to annot
# Here, 150 in mutate() function is the mean read length. I pulled out this number from multiqc results:
# /data/workspaces/lag/workspaces/lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/multiqc/multiqc/Alignment_QC_Report_data.pdf
countdata = annot %>%
            dplyr::select(GeneID, Length) %>%
            inner_join(countdata) %>%
            mutate(Length = Length-150+1) %>%
            dplyr::filter(Length>1) # filter short genes

# Now we divide first by the gene Length and then the colsums
# This changes all numeric colums
# The dot refers to the current value
tpm = countdata %>%
      mutate_if(is.numeric,funs(rpk=./Length*1e3)) %>%
      mutate_if(is.numeric,funs(tpm=./sum(.)*1e6)) %>%
      # only want to keep the new columns
      dplyr::select(GeneID, contains("rpk_tpm"), -Length_rpk_tpm)

#change column names back
names(tpm) = colnames(countdata)[-2]

```

# Check how comparable the TPMs are across samples

```{r checkpoint_plot1}

# Avoid having log(0) by simply adding the minimum TPM > 0 to each sample

expMin = apply(tpm[,-1],2,function(x){min(x[x>0])})

log_tpm = log10(t(tpm[,-1]) + expMin)

# Transposing with t() made TPM a matrix,
# Transpose back and revert to data.frame
# Make long format

tpm_long = data.frame(t(log_tpm)) %>%
           gather(key= "SRR", value = "log10_TPM" )

head(tpm_long)

ggplot(tpm_long,aes(x=SRR,y=log10_TPM)) +
       geom_boxplot() +
       coord_flip() +
       ylab(expression(log[10](TPM)))

ggsave(paste0(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/diffexpression/MW_20250528/results/TPM_comparison_boxplots_v2.pdf"), width = 12, height = 20, units = "cm")
#ggsave("/data/workspaces/lag/workspaces/lg-scrna/working_data/gokberk/setbp1/results/TPM_comparison_boxplots.pdf", width = 12, height = 20, units = "cm")
```

# Continue filtering before running DE analysis.

```{r prep4deseq_part2}

nrow(countdata)
# There are still 38,048 genes in our count table
# Filter further to 1) get rid off zeros in the matrix and
# improve normalization and 2) to filter out very lowly
# expressed genes, cuz you can't detect differential
# expression for those genes anyway.

# Calculate how many genes are expressed in each sample
tpm %>% summarise_if(is.numeric, funs(sum(.>=1)))
number_genes_expressed = tpm %>% summarise_if(is.numeric, funs(sum(.>=1)))

write.csv(number_genes_expressed, file =paste0(deseq_outDir, "number_genes_expressed.csv"))

# Move geneIDs to rownames, remove that column
rownames(tpm) = tpm[,1]
tpm = tpm[,-1]

# Condition only 10 samples are allowed to have a TPM < 1
# Note that this 10 samples threshold is arbitrary.

cut = dim(tpm)[2] - 10
sum(apply(tpm, 1, function(x){sum(x>=1)>=cut}))
# 13,116 genes pass this threshold
# So we reduced gene number from 38k to ~13k, success!

filt = tpm[apply(tpm, 1, function(x){sum(x>=1)>=cut}),]
tpm_filt_counts = rcounts[match(rownames(filt),rownames(rcounts)),]
write.csv(tpm_filt_counts, file =paste0(deseq_outDir, "tpm_filt_counts.csv"))

```

# Run differential expression analysis with "DESeq2".

```{r DEanalysis}

head(inf)

# Make sure that the IDs in the info table have
# the same order as the columns in the count table
rownames(inf) = inf$ID # make ID rownames
colnames(rcounts) = str_replace(colnames(rcounts), ".bam", "") # remove ".bam" from colnames
colnames(tpm_filt_counts) = str_replace(colnames(tpm_filt_counts), ".bam", "") # remove ".bam" from colnames
tpm_filt_counts = tpm_filt_counts[, rownames(inf)] # order columns based on sample list in inf

colnames(tpm_filt_counts) == rownames(inf) # check if it worked (all should be TRUE, yes)
colnames(tpm_filt_counts) = inf$ID # make IDs in info as colnames of filtered counts, data still the same
rownames(inf) = inf$ID # make IDs back to rowname
colnames(tpm_filt_counts) == rownames(inf) # check if it worked (all should be TRUE, yes, hope everything is fine)

nrow(tpm_filt_counts)
#[1] 13116

# DESeq design w/condition, sex, batch
dds = DESeqDataSetFromMatrix(countData = tpm_filt_counts,
                             colData = inf,
                             design = ~sex+batch+age_at_sampling+condition)

design(dds)
#~sex + batch + age_at_sampling + condition

dds = estimateSizeFactors(dds)
summary(sizeFactors(dds))

#saveRDS(dds, file="/data/workspaces/lag/workspaces/lg-scrna/working_data/gokberk/setbp1/results/DESeq/deseq_results/dds_v2.rds")
#readRDS(dds, file="/data/workspaces/lag/workspaces/lg-scrna/working_data/gokberk/setbp1/results/DESeq/deseq_results/dds_v2.rds")
saveRDS(dds, file =paste0(deseq_outDir,"/deseq_results/dds_v3.rds"))
#readRDS(dds, file =paste0(deseq_outDir,"/deseq_results/dds_v3.rds"))


vsd = varianceStabilizingTransformation(dds)

# Remove batch effect only for visualization purpose.
# Important Note: DESeq2 takes raw count numbers as input, and
# models batch effect. Thus, DE gene lists are not affected
# by this removeBatchEffect step.
# See what Mike Love says here: https://support.bioconductor.org/p/104290/
assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$batch)
plotPCA(vsd, "batch")

# Extract a matrix of normalized counts from the S4 object vsd
vsdMat = assay(vsd)

pdf(paste0(deseq_outDir, "/SDplots/tpm.pdf"), width = 7, height = 8)
meanSdPlot(as.matrix(tpm))
dev.off()

pdf(paste0(deseq_outDir, "/SDplots/log_tpm.pdf"), width = 7, height = 8)
meanSdPlot(t(log_tpm))
dev.off()

pdf(paste0(deseq_outDir, "/SDplots/vsdMat.pdf"), width = 7, height = 8)
meanSdPlot(vsdMat)
dev.off()

# PCA using the 500 most variable genes from the variance-stabilized data
nudge <- position_nudge(x = 2, y = 2)

p1 = plotPCA.san(vsd, intgroup = "condition", ntop = 500) +
  ggplot2::theme(legend.position = "top") #+
  #geom_text_repel(aes(label = name), size = 2) #+ 
#scale_fill_manual(values = c("purple","red", "green","orange"))

p2 = plotPCA.san(vsd, intgroup = "batch", ntop = 500) +
  ggplot2::theme(legend.position = "top") #+
  #geom_text(aes(label = name), position = nudge)
  #geom_text_repel(aes(label = name), size = 2)

p3 = plotPCA.san(vsd, intgroup = "sex", ntop = 500) +
  ggplot2::theme(legend.position = "top") #+
  #geom_text(aes(label = name), position = nudge)
  #geom_text_repel(aes(label = name), size = 2)

p4 = plotPCA.san(vsd, intgroup = "age_at_sampling", ntop = 500) +
  ggplot2::theme(legend.position = "top") #+
  #geom_text(aes(label = name), position = nudge)
  #geom_text_repel(aes(label = name), size = 2)

plot_grid(p1, p4, p3, p2, nrow = 2, ncol = 2)
ggsave(paste0(deseq_outDir, "/PCA/PCA_4categories_v3_noLabel.pdf"), width = 15, height = 10)

p1 = plotPCA.san(vsd, intgroup = "condition", ntop = 500) +
  ggplot2::theme(legend.position = "top") +
 geom_text_repel(aes(label = name), size = 2) + 
scale_fill_manual(values = c("purple","red", "green","orange"))

p2 = plotPCA.san(vsd, intgroup = "batch", ntop = 500) +
  ggplot2::theme(legend.position = "top") +
#  geom_text(aes(label = name), position = nudge) +
  geom_text_repel(aes(label = name), size = 2)

p3 = plotPCA.san(vsd, intgroup = "sex", ntop = 500) +
  ggplot2::theme(legend.position = "top") +
#  geom_text(aes(label = name), position = nudge) +
  geom_text_repel(aes(label = name), size = 2)
  
p4 = plotPCA.san(vsd, intgroup = "age_at_sampling", ntop = 500) +
  ggplot2::theme(legend.position = "top") +
#  geom_text(aes(label = name), position = nudge) +
  geom_text_repel(aes(label = name), size = 2)

plot_grid(p1, p4, p3, p2, nrow = 2, ncol = 2)
ggsave(paste0(deseq_outDir, "/PCA/PCA_4categories_v3_label.pdf"), width = 15, height = 10)


# Run DESeq for different combinations of conditions
# to make volcano plots
dds2 = DESeq(dds, betaPrior = FALSE)
saveRDS(dds2, file =paste0(deseq_outDir,"/deseq_results/dds2.rds"))
#readRDS(dds2, file =paste0(deseq_outDir,"/deseq_results/dds2.rds"))

# Control vs. Outside_degron
res1 = results(dds2, contrast = c("condition", "Outside_degron", "Control"))
res1 = lfcShrink(dds2, contrast = c("condition", "Outside_degron", "Control"), res=res1, type = "normal")
res1$symbol = mapIds(org.Hs.eg.db, keys=row.names(res1), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res1), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsOutsideDegron.csv"), quote = F)
#removeNA
res1_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_ControlvsOutsideDegron.csv"), header = T)
res1_deseq_results_noNA = na.omit(res1_deseq_results)
nrow(res1_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res1_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsOutsideDegron_noNA.csv"), quote = F)


# Control vs. within_degron/SGS
res2 = results(dds2, contrast = c("condition", "within_degron/SGS","Control"))
res2 = lfcShrink(dds2, contrast = c("condition", "within_degron/SGS","Control"), res=res2, type = "normal")
res2$symbol = mapIds(org.Hs.eg.db, keys=row.names(res2), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res2), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsWithin_degronSGS.csv"), quote = F)

#removeNA
res2_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_ControlvsWithin_degronSGS.csv"), header = T)
res2_deseq_results_noNA = na.omit(res2_deseq_results)
nrow(res2_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res2_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsWithin_degronSGS_noNA.csv"), quote = F)


# Control vs. Truncating
res3 = results(dds2, contrast = c("condition", "truncating", "Control"))
res3 = lfcShrink(dds2, contrast = c("condition", "truncating", "Control"), res=res3, type = "normal")
res3$symbol = mapIds(org.Hs.eg.db, keys=row.names(res3), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res3), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsTruncating.csv"), quote = F)

#removeNA
res3_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_ControlvsTruncating.csv"), header = T)
res3_deseq_results_noNA = na.omit(res3_deseq_results)
nrow(res3_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res3_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_ControlvsTruncating_noNA.csv"), quote = F)


# Control vs. Outside
#res4 = results(dds2, contrast = c("condition", "Control", "Outside"))
#res4 = lfcShrink(dds2, contrast = c("condition", "Control", "Outside"), res=res4, type = "normal")
#res4$symbol = mapIds(org.Hs.eg.db, keys=row.names(res4), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
#write.csv(as.data.frame(res4), paste0(deseq_outDir,
#                                     "/deseq_results/res_ControlvsOutside.csv"), quote = F)

# Outside_degron vs. within_degron/SGS
res5 = results(dds2, contrast = c("condition", "Outside_degron", "within_degron/SGS"))
res5 = lfcShrink(dds2, contrast = c("condition", "Outside_degron", "within_degron/SGS"), res=res5, type = "normal")
res5$symbol = mapIds(org.Hs.eg.db, keys=row.names(res5), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res5), paste0(deseq_outDir,
                                     "/deseq_results/res_Outside_degronvsWithin_degronSGS.csv"), quote = F)

#removeNA
res5_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_Outside_degronvsWithin_degronSGS.csv"), header = T)
res5_deseq_results_noNA = na.omit(res5_deseq_results)
nrow(res5_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res5_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_Outside_degronvsWithin_degronSGS_noNA.csv"), quote = F)


# within_degron/SGS vs. truncating
res6 = results(dds2, contrast = c("condition", "within_degron/SGS", "truncating"))
res6 = lfcShrink(dds2, contrast = c("condition", "within_degron/SGS", "truncating"), res=res6, type = "normal")
res6$symbol = mapIds(org.Hs.eg.db, keys=row.names(res6), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res6), paste0(deseq_outDir,
                                     "/deseq_results/res_within_degronSGSvstruncating.csv"), quote = F)

#removeNA
res6_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_within_degronSGSvstruncating.csv"), header = T)
res6_deseq_results_noNA = na.omit(res6_deseq_results)
nrow(res6_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res6_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_within_degronSGSvstruncating_noNA.csv"), quote = F)

# Outside_degron vs. truncating
res7 = results(dds2, contrast = c("condition", "Outside_degron", "truncating"))
res7 = lfcShrink(dds2, contrast = c("condition", "Outside_degron", "truncating"), res=res7, type = "normal")
res7$symbol = mapIds(org.Hs.eg.db, keys=row.names(res7), column="SYMBOL", keytype="ENSEMBL", multiVals="first")
write.csv(as.data.frame(res7), paste0(deseq_outDir,
                                     "/deseq_results/res_Outside_degronvstruncating.csv"), quote = F)

#removeNA
res7_deseq_results = read.csv(paste0(deseq_outDir, "/deseq_results/res_Outside_degronvstruncating.csv"), header = T)
res7_deseq_results_noNA = na.omit(res7_deseq_results)
nrow(res7_deseq_results_noNA)
#[1]  11819
write.csv(as.data.frame(res7_deseq_results_noNA), paste0(deseq_outDir,
                                     "/deseq_results/res_Outside_degronvstruncating_noNA.csv"), quote = F)

```

```{r volcanoPlots}

# Plot and save volcanoplots

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-ControlvsOutsideDegron.pdf"), width = 7, height = 8)
EnhancedVolcano(res1, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-ControlvsWithin_degronSGS.pdf"), width = 7, height = 8)
EnhancedVolcano(res2, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-ControlvsTruncating.pdf"), width = 7, height = 8)
EnhancedVolcano(res3, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()

#pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-ControlvsOutside.pdf"), width = 7, height = 8)
#EnhancedVolcano(res4, lab = NA, x="log2FoldChange", y="padj",
#                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
#                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
#                gridlines.minor = FALSE, gridlines.major = FALSE, )
#dev.off()

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-Outside_degronvsWithin_degronSGS.pdf"), width = 7, height = 8)
EnhancedVolcano(res5, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-within_degronSGSvstruncating.pdf"), width = 7, height = 8)
EnhancedVolcano(res6, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()

pdf(paste0(deseq_outDir, "volcanoPlots/Volcano-Outside_degronvstruncating.pdf"), width = 7, height = 8)
EnhancedVolcano(res7, lab = NA, x="log2FoldChange", y="padj",
                xlim = c(-2, 2), ylim = c(0,15), captionLabSize = 9,
                col=c('grey75', 'grey50', 'grey25', 'red','blue'), pCutoff = 0.1, FCcutoff = 0.6,
                gridlines.minor = FALSE, gridlines.major = FALSE)
dev.off()


```

```{r getDEgeneCounts}

get_upregulated = function(df){

    df = as.data.frame(df)
  
    key = intersect(rownames(df)[which(df$log2FoldChange>=1)], rownames(df)[which(df$padj<=0.05)])

    results = as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

get_downregulated = function(df){

    df = as.data.frame(df)
  
    key = intersect(rownames(df)[which(df$log2FoldChange<=-1)], rownames(df)[which(df$padj<=0.05)])

    results = as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

for (i in 1:7) {
  if (i==4) next # skip ControlvsOutside
  up_list = get_upregulated(get(paste0("res", i)))
  write.table(up_list, paste0(deseq_outDir,"/sigDEgeneLists/sigUpRegulatedGenes_res", i, ".tab"),
              quote = F, sep = "\t")
  
  }

for (i in 1:7) {
  if (i==4) next # skip ControlvsOutside
  down_list = get_downregulated(get(paste0("res", i)))
  write.table(down_list, paste0(deseq_outDir,"/sigDEgeneLists/sigDownRegulatedGenes_res", i, ".tab"),
              quote = F, sep = "\t")
  
  }
#check number of up- and down- regulated genes for each comparison

for (i in 1:7) {
  if (i==4) next # skip ControlvsOutside
  summary_list = summary(get(paste0("res", i)))}

  #write.table(summary_list, paste0(deseq_outDir,"/sigDEgeneLists/summary_res", i, ".tab"),
 #             quote = F, sep = "\t")} #don't know why can't put in a table

outside_in_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res5.tab"), header = T)
outside_in_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res5.tab"), header = T)

inside_truncating_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res6.tab"), header = T)
inside_truncating_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res6.tab"), header = T)

outside_truncating_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res7.tab"), header = T)
outside_truncating_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res7.tab"), header = T)

```


```{r heatmaps}

# We can’t plot the entire data set, let’s just select the top 300 by padj
# Get top 300 differentially expressed genes between Control and Outside_degron samples

# First, add gene IDs as a column
res1$geneIDs = rownames(res1)

sigGenes_ControlvsOutside_degron = as.data.frame(res1) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

# We’ll use normalised expression values, so we’ll use the vst function -> Changed this, see below
#plotDat_ControlvsOutside_degron <- vst(dds)[sigGenes_ControlvsOutside_degron,] %>% 
#  assay()
# Here we used Batch Effect Removed (with limma) vsd file to see heatmaps without batch effect
plotDat_ControlvsOutside_degron <- vsd[sigGenes_ControlvsOutside_degron,] %>% 
  assay()

z.mat_ControlvsOutside_degron <- t(scale(t(plotDat_ControlvsOutside_degron), center=TRUE, scale=TRUE))

myPalette <- c("royalblue3", "ivory", "orangered3")
myRamp <- colorRamp2(c(-2, 0, 2), myPalette)

ha = HeatmapAnnotation(df = colData(dds)[,c("condition", "batch", "sex")],
                        col = list(condition = c("Control" = "pink", "Outside" = "purple",
                                                 "Outside_degron"="lightblue", "truncating" = "red",
                                                 "within_degron/SGS" = "green"),
                                   batch = c("b1" = "lightblue", "b2" = "darkblue"),
                                   sex = c("M" = "orange", "F" = "darkgreen")))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_ControlvsOutside_degron_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_ControlvsOutside_degron, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

# Get top 300 differentially expressed genes between Control and within_degron/SGS

res2$geneIDs = rownames(res2)

sigGenes_ControlvsWithin_degronSGS = as.data.frame(res2) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

#plotDat_ControlvsWithin_degronSGS <- vst(dds)[sigGenes_ControlvsWithin_degronSGS,] %>% 
#  assay()
plotDat_ControlvsWithin_degronSGS <- vsd[sigGenes_ControlvsWithin_degronSGS,] %>% 
  assay()
z.mat_ControlvsWithin_degronSGS <- t(scale(t(plotDat_ControlvsWithin_degronSGS), center=TRUE, scale=TRUE))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_ControlvsWithin_degronSGS_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_ControlvsWithin_degronSGS, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

# Get top 300 differentially expressed genes between Control vs. Truncating

res3$geneIDs = rownames(res3)

sigGenes_ControlvsTruncating = as.data.frame(res3) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

#plotDat_ControlvsTruncating <- vst(dds)[sigGenes_ControlvsTruncating,] %>% 
#  assay()
plotDat_ControlvsTruncating <- vsd[sigGenes_ControlvsTruncating,] %>% 
  assay()
z.mat_ControlvsTruncating <- t(scale(t(plotDat_ControlvsTruncating), center=TRUE, scale=TRUE))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_ControlvsTruncating_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_ControlvsTruncating, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

# Get top 300 differentially expressed genes between Control vs. Outside

#res4$geneIDs = rownames(res4)

#sigGenes_ControlvsOutside = as.data.frame(res4) %>% 
#                                   top_n(300, wt=-padj) %>% 
#                                   pull("geneIDs")

#plotDat_ControlvsOutside <- vst(dds)[sigGenes_ControlvsOutside,] %>% 
#  assay()
#plotDat_ControlvsOutside <- vsd[sigGenes_ControlvsOutside,] %>% 
#  assay()
#z.mat_ControlvsOutside <- t(scale(t(plotDat_ControlvsOutside), center=TRUE, scale=TRUE))

#pdf(file = paste0(deseq_outDir, "heatmap_top300DEgenes_ControlvsOutside_batchEffRemoved.pdf"),
#    width = 12, height = 8)
#Heatmap(z.mat_ControlvsOutside, name = "z-score",
#        col = myRamp,            
#        show_row_name = FALSE,
#        #split=3,
#        rect_gp = gpar(col = "lightgrey", lwd=0.3),
#        top_annotation = ha)
#dev.off()

# Get top 300 differentially expressed genes between Outside_degron vs. within_degron/SGS

res5$geneIDs = rownames(res5)

sigGenes_Outside_degronvsWithin_degronSGS = as.data.frame(res5) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

#plotDat_Outside_degronvsWithin_degronSGS <- vst(dds)[sigGenes_Outside_degronvsWithin_degronSGS,] %>% 
#  assay()
plotDat_Outside_degronvsWithin_degronSGS <- vsd[sigGenes_Outside_degronvsWithin_degronSGS,] %>% 
  assay()
z.mat_Outside_degronvsWithin_degronSGS <- t(scale(t(plotDat_Outside_degronvsWithin_degronSGS), center=TRUE, scale=TRUE))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_Outside_degronvsWithin_degronSGS_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_Outside_degronvsWithin_degronSGS, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

# Get top 300 differentially expressed genes between within_degron/SGS vs. truncating

res6$geneIDs = rownames(res6)

sigGenes_within_degronSGSvstruncating = as.data.frame(res6) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

#plotDat_Outside_degronvsWithin_degronSGS <- vst(dds)[sigGenes_within_degronSGSvstruncating,] %>% 
#  assay()
plotDat_within_degronSGSvstruncating <- vsd[sigGenes_within_degronSGSvstruncating,] %>% 
  assay()
z.mat_within_degronSGSvstruncating <- t(scale(t(plotDat_within_degronSGSvstruncating), center=TRUE, scale=TRUE))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_within_degronSGSvstruncating_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_within_degronSGSvstruncating, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

# Get top 300 differentially expressed genes between Outside_degron vs. truncating

res7$geneIDs = rownames(res7)

sigGenes_Outside_degronvstruncating = as.data.frame(res7) %>% 
                                   top_n(300, wt=-padj) %>% 
                                   pull("geneIDs")

#plotDat_Outside_degronvsWithin_degronSGS <- vst(dds)[sigGenes_Outside_degronvsWithin_degronSGS,] %>% 
#  assay()
plotDat_Outside_degronvstruncating <- vsd[sigGenes_Outside_degronvstruncating,] %>% 
  assay()
z.mat_Outside_degronvstruncating <- t(scale(t(plotDat_Outside_degronvstruncating), center=TRUE, scale=TRUE))

pdf(file = paste0(deseq_outDir, "/heatmaps/heatmap_top300DEgenes_Outside_degronvstruncating_batchEffRemoved.pdf"),
    width = 12, height = 8)
Heatmap(z.mat_Outside_degronvstruncating, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        #split=3,
        rect_gp = gpar(col = "lightgrey", lwd=0.3),
        top_annotation = ha)
dev.off()

```

```{r goanalysis}
# only for upregulated genes
for (i in 1:7) {
      
  if (i==4) next
  tmp_df = data.frame(rn = rownames(get(paste0("res",i))), get(paste0("res",i))) %>%
       na.omit() %>%
       mutate(up = ifelse(log2FoldChange>0, padj, 1))

# set the p-value for down regulations to 1
# make a named vector
tmp_condition_up = tmp_df$up
names(tmp_condition_up) <- tmp_df$rn

topDiffGenes<-function(x){ return(x<=0.1) }

# 1) BP

tg.BP = new("topGOdata",#just a name
            description="tmp_condition_up",
            ontology="BP",
            allGenes=tmp_condition_up, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.BP, paste0(deseq_outDir, "GOtermAnalysis/GO_BP_upRegulated_res", i,".rds"))

numGenes(tg.BP)

numSigGenes(tg.BP)

fisher.res = runTest(tg.BP, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_BP = GenTable(tg.BP, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])

# 2) CC

tg.CC = new("topGOdata",#just a name
            description="tmp_condition_up",
            ontology="CC",
            allGenes=tmp_condition_up, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.CC, paste0(deseq_outDir, "GOtermAnalysis/GO_CC_upRegulated_res", i,".rds"))

numGenes(tg.CC)

numSigGenes(tg.CC)

fisher.res = runTest(tg.CC, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_CC = GenTable(tg.CC, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])

# 3) MF

tg.MF = new("topGOdata",#just a name
            description="tmp_condition_up",
            ontology="MF",
            allGenes=tmp_condition_up, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.MF, paste0(deseq_outDir, "GOtermAnalysis/GO_MF_upRegulated_res", i,".rds"))

numGenes(tg.MF)

numSigGenes(tg.MF)

fisher.res = runTest(tg.MF, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_MF = GenTable(tg.MF, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])

# Plot

fishtab_BP2 = fishtab_BP[1:10, c(2,4)]
fishtab_BP2$ontology = "BP"
fishtab_CC2 = fishtab_CC[1:10, c(2,4)]
fishtab_CC2$ontology = "CC"
fishtab_MF2 = fishtab_MF[1:10, c(2,4)]
fishtab_MF2$ontology = "MF"

p1 = ggplot(fishtab_BP2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "green") + ylim(0, max(fishtab_CC2$Significant)) +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() + 
  theme(axis.title.x=element_blank(),
        axis.title.y=element_blank())
p2 = ggplot(fishtab_CC2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "red") + labs(x = "GO Terms") +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() +
  theme(axis.title.x=element_blank())
p3 = ggplot(fishtab_MF2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "blue") +  labs(y = "Number of genes") +  ylim(0, max(fishtab_CC2$Significant)) +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() + 
  theme(axis.title.y=element_blank())

plot_grid(p1, p2, p3, nrow = 3, ncol = 1)
ggsave(paste0(deseq_outDir, "GOtermAnalysis/GOtermsUp_res", i,".pdf"), width = 8, height = 15)

go_termsList_BP = fishtab_BP[1:10,] %>% arrange(desc(Significant))
write.table(go_termsList_BP, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_UpBP_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")


go_termsList_CC = fishtab_CC[1:10, ]%>% arrange(desc(Significant))
write.table(go_termsList_CC, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_UpCC_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")


go_termsList_MF = fishtab_MF[1:10, ]%>% arrange(desc(Significant))
write.table(go_termsList_MF, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_UpMF_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")
}


# for downregulated genes

for (i in 1:7) {
      
  if (i==4) next
  tmp_df = data.frame(rn = rownames(get(paste0("res",i))), get(paste0("res",i))) %>%
       na.omit() %>%
       mutate(down = ifelse(log2FoldChange<0, padj, 1))

# set the p-value for down regulations to 1
# make a named vector
tmp_condition_down = tmp_df$down
names(tmp_condition_down) <- tmp_df$rn

topDiffGenes<-function(x){ return(x<=0.1) }

# 1) BP

tg.BP = new("topGOdata",#just a name
            description="tmp_condition_down",
            ontology="BP",
            allGenes=tmp_condition_down, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.BP, paste0(deseq_outDir, "GOtermAnalysis/GO_BP_downRegulated_res", i,".rds"))

numGenes(tg.BP)

numSigGenes(tg.BP)

fisher.res = runTest(tg.BP, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_BP = GenTable(tg.BP, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])
# retrieve geens2GO list from the annoation in GO data
  allGO_BP = genesInTerm(tg.BP)

# 2) CC

tg.CC = new("topGOdata",#just a name
            description="tmp_condition_down",
            ontology="CC",
            allGenes=tmp_condition_down, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.CC, paste0(deseq_outDir, "GOtermAnalysis/GO_CC_downRegulated_res", i,".rds"))

numGenes(tg.CC)

numSigGenes(tg.CC)

fisher.res = runTest(tg.CC, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_CC = GenTable(tg.CC, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])

# retrieve geens2GO list from the annoation in GO data
  allGO_CC = genesInTerm(tg.CC)

# 3) MF

tg.MF = new("topGOdata",#just a name
            description="tmp_condition_down",
            ontology="MF",
            allGenes=tmp_condition_down, #named vector with adj. p-values
            geneSel=topDiffGenes,
            annot=annFUN.org,
            nodeSize = 10, #minimum number of genes in a GO categorie
            ID="ENSEMBL",
            mapping="org.Hs.eg.db")

saveRDS(tg.MF, paste0(deseq_outDir, "GOtermAnalysis/GO_MF_downRegulated_res", i,".rds"))

numGenes(tg.MF)

numSigGenes(tg.MF)

fisher.res = runTest(tg.MF, algorithm="elim", statistic="fisher")
fisher.res

#p-value adjustment
qval = p.adjust(score(fisher.res), method = "BH")
summary(qval<0.1)

fishtab_MF = GenTable(tg.MF, Fisher=fisher.res, orderBy="Fisher", topNodes = geneData(fisher.res)[4])

# retrieve geens2GO list from the annoation in GO data
  allGO_MF = genesInTerm(tg.MF)
  
# Plot

fishtab_BP2 = fishtab_BP[1:10, c(2,4)]
fishtab_BP2$ontology = "BP"
fishtab_CC2 = fishtab_CC[1:10, c(2,4)]
fishtab_CC2$ontology = "CC"
fishtab_MF2 = fishtab_MF[1:10, c(2,4)]
fishtab_MF2$ontology = "MF"

p1 = ggplot(fishtab_BP2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "green") + ylim(0, max(fishtab_CC2$Significant)) +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() + 
  theme(axis.title.x=element_blank(),
        axis.title.y=element_blank())
p2 = ggplot(fishtab_CC2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "red") + labs(x = "GO Terms") +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() +
  theme(axis.title.x=element_blank())
p3 = ggplot(fishtab_MF2, aes(x = reorder(Term, Significant), y = Significant)) +
  geom_bar(stat = "identity", fill = "blue") +  labs(y = "Number of genes") +  ylim(0, max(fishtab_CC2$Significant)) +
  coord_flip() + facet_wrap(~ontology) + theme_minimal() + 
  theme(axis.title.y=element_blank())

plot_grid(p1, p2, p3, nrow = 3, ncol = 1)
ggsave(paste0(deseq_outDir, "GOtermAnalysis/GOtermsDown_res", i,".pdf"), width = 8, height = 15)

go_termsList_BP = fishtab_BP[1:10,] %>% arrange(desc(Significant))
write.table(go_termsList_BP, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_downBP_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")


go_termsList_CC = fishtab_CC[1:10, ]%>% arrange(desc(Significant))
write.table(go_termsList_CC, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_downCC_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")


go_termsList_MF = fishtab_MF[1:10, ]%>% arrange(desc(Significant))
write.table(go_termsList_MF, 
              paste0(deseq_outDir, "GOtermAnalysis/GO_termList_downMF_res_", i, ".txt"), quote = F, row.names = F, sep = "\t")
}



```



```{r vennDiagram}
#deseq_outDir = paste(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_HDF/diffexpression/MW_20221018/results/DESeq/",sep="")


# Outside degron vs within degron vs truncating

# Significantly UPregulated genes in Control vs 3 conditions
outside_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res1.tab"), header = T)
inside_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res2.tab"), header = T)
truncating_up = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigUpRegulatedGenes_res3.tab"), header = T)

n12_up = length(intersect(rownames(outside_up),rownames(inside_up)))
write.table(outside_up[intersect(rownames(outside_up),rownames(inside_up)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/sigUpRegulatedGenes_outsidevsInsideOverlap_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n23_up = length(intersect(rownames(inside_up),rownames(truncating_up)))
write.table(outside_up[intersect(rownames(inside_up),rownames(truncating_up)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/sigUpRegulatedGenes_insidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n13_up = length(intersect(rownames(outside_up),rownames(truncating_up)))
write.table(outside_up[intersect(rownames(outside_up),rownames(truncating_up)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/sigUpRegulatedGenes_outsidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n123_up = length(intersect(rownames(truncating_up),intersect(rownames(outside_up),rownames(inside_up))))
write.table(outside_up[intersect(rownames(truncating_up),intersect(rownames(outside_up),rownames(inside_up))),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/sigUpRegulatedGenes_truncatingvsOutsidevsInside_geneList.tab"),
            quote = F, row.names = F, col.names = F)

venn_up = draw.triple.venn(nrow(outside_up),nrow(inside_up),nrow(truncating_up),
                 n12_up, n23_up, n13_up, n123_up,
                 c("Outside_degron", "Inside_degron", "Truncating"),
                 fill = c("blue", "red", "green"))
pdf(paste0(deseq_outDir, "vennDiagrams/venn_sigDEgenesUp_outsideVSinsideVStruncating.pdf"),
    width = 12, height = 8)
grid.draw(venn_up)
dev.off()

# Significantly DOWNregulated genes in Control vs 3 conditions
outside_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res1.tab"), header = T)
inside_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res2.tab"), header = T)
truncating_down = read.table(paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_res3.tab"), header = T)


n12_down = length(intersect(rownames(outside_down),rownames(inside_down)))
write.table(outside_down[intersect(rownames(outside_down),rownames(inside_down)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_outsidevsInsideOverlap_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n23_down = length(intersect(rownames(inside_down),rownames(truncating_down)))
write.table(outside_down[intersect(rownames(inside_down),rownames(truncating_down)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_insidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n13_down = length(intersect(rownames(outside_down),rownames(truncating_down)))
write.table(outside_down[intersect(rownames(outside_down),rownames(truncating_down)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_outsidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n123_down = length(intersect(rownames(truncating_down),intersect(rownames(outside_down),rownames(inside_down))))
write.table(outside_down[intersect(rownames(truncating_down),intersect(rownames(outside_down),rownames(inside_down))),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDownRegulatedGenes_truncatingvsOutsidevsInside_geneList.tab"),
            quote = F, row.names = F, col.names = F)

venn_down = draw.triple.venn(nrow(outside_down),nrow(inside_down),nrow(truncating_down),
                 n12_down, n23_down, n13_down, n123_down,
                 c("Outside_degron", "Inside_degron", "Truncating"),
                 fill = c("blue", "red", "green"))
pdf(paste0(deseq_outDir, "vennDiagrams/venn_sigDEgenesDown_outsideVSinsideVStruncating.pdf"),
    width = 12, height = 8)
grid.draw(venn_down)
dev.off()

# Significantly DEGs in Control vs 3 conditions
outside_DEG = read.csv(paste0(deseq_outDir, "deseq_results/res_ControlvsOutsideDegron.csv"), header = T)
outside_DEG = subset(outside_DEG, outside_DEG$padj<0.1)
outside_DEG = subset(outside_DEG, outside_DEG$log2FoldChange >= 1 | outside_DEG$log2FoldChange <= -1)
outside_DEG = data.frame(outside_DEG)
rownames(outside_DEG)=outside_DEG$X
outside_DEG = outside_DEG[,-1]
nrow(outside_DEG)
#[1] 208
write.table(outside_DEG, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsOutside.tab"),
              quote = F, sep = "\t")
#removeNA
outside_DEG = read.table(paste0(deseq_outDir, "/sigDEgeneLists/filteredDEG_ControlvsOutside.tab"), header = T)
outside_DEG_noNA = na.omit(outside_DEG)
nrow(outside_DEG_noNA)
#[1] 197
write.table(outside_DEG_noNA, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsOutside_noNA.tab"),
              quote = F, sep = "\t")



inside_DEG = read.csv(paste0(deseq_outDir, "deseq_results/res_ControlvsWithin_degronSGS.csv"), header = T)
inside_DEG = subset(inside_DEG, inside_DEG$padj<0.1)
inside_DEG = subset(inside_DEG, inside_DEG$log2FoldChange >= 1 | inside_DEG$log2FoldChange <= -1)
inside_DEG = data.frame(inside_DEG)
rownames(inside_DEG)=inside_DEG$X
inside_DEG = inside_DEG[,-1]
nrow(inside_DEG)
#[1] 389
write.table(inside_DEG, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsWithin_degronSGS.tab"),
              quote = F, sep = "\t")
#removeNA
inside_DEG = read.table(paste0(deseq_outDir, "/sigDEgeneLists/filteredDEG_ControlvsWithin_degronSGS.tab"), header = T)
inside_DEG_noNA = na.omit(inside_DEG)
nrow(inside_DEG_noNA)
#[1] 366
write.table(inside_DEG_noNA, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsWithin_degronSGS_noNA.tab"),
              quote = F, sep = "\t")


truncating_DEG = read.csv(paste0(deseq_outDir, "deseq_results/res_ControlvsTruncating.csv"), header = T)
truncating_DEG = subset(truncating_DEG, truncating_DEG$padj<0.1)
truncating_DEG = subset(truncating_DEG, truncating_DEG$log2FoldChange >= 1 | truncating_DEG$log2FoldChange <= -1)
truncating_DEG = data.frame(truncating_DEG)
rownames(truncating_DEG)=truncating_DEG$X
truncating_DEG = truncating_DEG[,-1]
nrow(truncating_DEG)
#[1] 367
write.table(truncating_DEG, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsTruncating.tab"),
              quote = F, sep = "\t")

#removeNA
truncating_DEG = read.table(paste0(deseq_outDir, "/sigDEgeneLists/filteredDEG_ControlvsTruncating.tab"), header = T)
truncating_DEG_noNA = na.omit(truncating_DEG)
nrow(truncating_DEG_noNA)
#[1] 345
write.table(truncating_DEG_noNA, paste0(deseq_outDir,"/sigDEgeneLists/filteredDEG_ControlvsTruncating_noNA.tab"),
              quote = F, sep = "\t")


n12_DEG = length(intersect(rownames(outside_DEG),rownames(inside_DEG)))
write.table(outside_DEG[intersect(rownames(outside_DEG),rownames(inside_DEG)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_outsidevsInsideOverlap_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n23_DEG = length(intersect(rownames(inside_DEG),rownames(truncating_DEG)))
write.table(outside_DEG[intersect(rownames(inside_DEG),rownames(truncating_DEG)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_insidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n13_DEG = length(intersect(rownames(outside_DEG),rownames(truncating_DEG)))
write.table(outside_DEG[intersect(rownames(outside_DEG),rownames(truncating_DEG)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_outsidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n123_DEG = length(intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))))
write.table(outside_DEG[intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_truncatingvsOutsidevsInside_geneList.tab"),
            quote = F, row.names = F, col.names = F)

venn_DEG = draw.triple.venn(nrow(outside_DEG),nrow(inside_DEG),nrow(truncating_DEG),
                 n12_DEG, n23_DEG, n13_DEG, n123_DEG,
                 c("Outside_degron", "Inside_degron", "Truncating"),
                 fill = c("blue", "red", "green"))
pdf(paste0(deseq_outDir, "vennDiagrams/venn_DEG_outsideVSinsideVStruncating.pdf"),
    width = 12, height = 8)
grid.draw(venn_DEG)
dev.off()

#for overlapping of DEGs, autism and ID genes-----------------here--> this is odd, do i need this?start here
#outside
outsideDEG_ASD = length(intersect(rownames(outside_DEG),rownames(autism_genes)))
outsideDEG_ID = length(intersect(rownames(outside_DEG),rownames(ID_genes)))

truncatingDEG_ASD = length(intersect(rownames(autism_genes),rownames(truncating_DEG)))
truncatingDEG_ID = length(intersect(rownames(ID_genes),rownames(truncating_DEG)))
# 0 overlap for ASD or ID genes with truncating only
insideDEG_ASD = length(intersect(rownames(inside_DEG),rownames(autism_genes)))
insideDEG_ID = length(intersect(rownames(inside_DEG),rownames(ID_genes)))
# 0 overlap for ASD or ID genes with inside only

115
write.table(outside_DEG[intersect(rownames(outside_DEG),rownames(truncating_DEG)),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_outsidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)
n123_DEG = length(intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))))
write.table(outside_DEG[intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))),"symbol"],
            paste0(deseq_outDir, "/sigDEgeneLists/DEG_truncatingvsOutsidevsInside_geneList.tab"),
            quote = F, row.names = F, col.names = F)

venn_DEG = draw.triple.venn(nrow(outside_DEG),nrow(inside_DEG),nrow(truncating_DEG),
                 n12_DEG, n23_DEG, n13_DEG, n123_DEG,
                 c("Outside_degron", "Inside_degron", "Truncating"),
                 fill = c("blue", "red", "green"))
pdf(paste0(deseq_outDir, "vennDiagrams/venn_DEG_outsideVSinsideVStruncating.pdf"),
    width = 12, height = 8)
grid.draw(venn_DEG)
dev.off()

# Make a Venn Diagram for all sig DE genes
# and also a print out list of them
# done

outside_DEG = rbind(outside_up, outside_down)
inside_DEG = rbind(inside_up, inside_down)
truncating_DEG = rbind(truncating_up, truncating_down)

n12_DEG = length(intersect(rownames(outside_DEG),rownames(inside_DEG)))
write.table(outside_DEG[intersect(rownames(outside_DEG),rownames(inside_DEG)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDEG_outsidevsInsideOverlap_geneList.tab"),
            quote = F, row.names = F, col.names = F)

n23_DEG = length(intersect(rownames(inside_DEG),rownames(truncating_DEG)))
write.table(outside_DEG[intersect(rownames(inside_DEG),rownames(truncating_DEG)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDEG_insidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)

n13_DEG = length(intersect(rownames(outside_DEG),rownames(truncating_DEG)))
write.table(outside_DEG[intersect(rownames(outside_DEG),rownames(truncating_DEG)),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDEG_outsidevsTruncating_geneList.tab"),
            quote = F, row.names = F, col.names = F)

n123_DEG = length(intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))))
write.table(outside_DEG[intersect(rownames(truncating_DEG),intersect(rownames(outside_DEG),rownames(inside_DEG))),"symbol"],
            paste0(deseq_outDir, "sigDEgeneLists/sigDEG_truncatingvsOutsidevsInside_geneList.tab"),
            quote = F, row.names = F, col.names = F)

venn_down = draw.triple.venn(nrow(outside_DEG),nrow(inside_DEG),nrow(truncating_DEG),
                 n12_DEG, n23_DEG, n13_DEG, n123_DEG,
                 c("Outside_degron", "Inside_degron", "Truncating"),
                 fill = c("blue", "red", "green"))
pdf(paste0(deseq_outDir, "vennDiagrams/venn_sigDEG_outsideVSinsideVStruncating.pdf"),
    width = 12, height = 8)
grid.draw(venn_down)
dev.off()

```

# RRHO for transcriptomic similarity between groups

```{r RRHO}
#https://systems.crump.ucla.edu/rankrank/index.php
# do for all genes
# make lists for RRHO and save them as csv files, this works
for (i in 1:7) {
  if (i==4) next
  
  tmp_dds = get(paste0("res", i))
  
  # remove genes which could not be evaluated
  # tmp_subset = data.frame(tmp_dds,
  #                         stringsAsFactors = F %>%
  #                         )
  
#  tmp_dds = tmp_dds[tmp_dds$padj<=0.05,] # here we get only sig DEs 
  tmp_list = data.frame(element = tmp_dds$symbol,
                        value = -log10(tmp_dds$pvalue)*sign(tmp_dds$log2FoldChange))
  
  write.csv(tmp_list, 
            file = paste0(deseq_outDir, "RRHO_allGenes/list_", i, ".csv"),
            quote = F, row.names = F)
  
}

# run RRHO for each list pair, only do for res 1, 2, 3: Res1: control vs outside, Res2: control vs within SGS, Res3: control vs truncating LoF
# now it also works
  for (j in  1:3) {
    if (j==4) next
    
    tmp_list1 = read.csv(paste0(deseq_outDir, "RRHO_allGenes/list_", j, ".csv"))
    tmp_list1 = na.omit(tmp_list1) #remove the rows without a gene symbol
    tmp_list1 = tmp_list1[order(tmp_list1$element, -abs(tmp_list1$value)),] #remove the duplicate genes and keep the one with a larger DE
    tmp_list1 = tmp_list1[!duplicated(tmp_list1$element),]
    
    for (k in  1:3) {
      if (k==4|k==j) next
      
      tmp_list2 = read.csv(paste0(deseq_outDir, "RRHO_allGenes/list_", k, ".csv"))
      tmp_list2 = na.omit(tmp_list2)
      tmp_list2 = tmp_list2[order(tmp_list2$element, -abs(tmp_list2$value)),]
      tmp_list2 = tmp_list2[!duplicated(tmp_list2$element),]
      
  RRHO=    RRHO(tmp_list1, tmp_list2, 
           stepsize = 100, 
           labels = (c(as.character(j),as.character(k))), 
           alternative = "enrichment", #100 us recommended as the step size
           plots = T, outputdir = paste0(deseq_outDir, "RRHO_allGenes/"), 
           BY = T,
           log10.ind = T) 
#same gene list length so no need to correct for that      

              }
  }



```


### this find overlaps between two gene lists, do fisher exact test & plot venn diagram
```{r overlaps}
# Load necessary libraries
# library(dplyr)
# library(VennDiagram)
# library(gridExtra)
# library(stats)  # for Fisher's exact test

# Define a function to plot Venn diagram and save as PDF
plot_venn_pdf <- function(list_of_genes, labels, output_file) {
  pdf(file = output_file, width = 10, height = 10)
  venn.plot <- venn.diagram(
    x = list_of_genes,
    category.names = labels,
    filename = NULL,  # Do not save as a file/.png
    output = TRUE,
    col = "black",
    fill = c("cornflowerblue", "green"),  # Adjusted to 2 colors for 2 sets
    alpha = 0.50,
    cex = 2,
    fontface = "bold",
    fontfamily = "sans",
    cat.cex = 2,
    cat.fontface = "bold",
    cat.default.pos = "outer",
    cat.pos = c(-27, 27),  # Adjusted for 2 sets
    cat.dist = c(0.055, 0.055),  # Adjusted for 2 sets
    cat.fontfamily = "sans",
    rotation.degree = 1  # Changed from rotation to rotation.degree
  )
  grid.draw(venn.plot)
  dev.off()
}


# Define a function to perform Fisher's exact test
perform_fishers_test <- function(set1, set2, universe_size) {
  overlap <- length(intersect(set1, set2))
  only_set1 <- length(set1) - overlap
  only_set2 <- length(set2) - overlap
  neither <- universe_size - length(union(set1, set2))
  
  contingency_table <- matrix(c(overlap, only_set1, only_set2, neither), 
                              nrow = 2, 
                              dimnames = list(c("In Set2", "Not in Set2"),
                                              c("In Set1", "Not in Set1")))
  
  fisher_test_result <- fisher.test(contingency_table)
  return(fisher_test_result$p.value)
}

# Read genes of interest
overlapList_Dir = paste(dir, "lg-scrna/working_data/Maggie/working/bulk_rna_SETBP1_iNeuron/",sep="")

setbp1_targets <- read.csv(paste0(overlapList_Dir, "Top70_putative_SETBP1_bindingsites.csv"), header = F)$V1
setbp1_targets[1] = "EP300"
ID_genes <- read.csv(paste0(overlapList_Dir, "ID_whole_panelapp.csv"), header = F)$V1
autism_genes <- read.csv2(paste0(overlapList_Dir, "autism_whole_panelapp.csv"), header = F)$V1
ectopicG870S <- read.csv2(paste0(overlapList_Dir, "ectopicG870S.csv"), header = F)$V1 #this one somehow has a lot of tabs in front of the gene name

# only process those if they are expressed
## expressed genes
universe <- res1_deseq_results_noNA$symbol
universe <- universe[!is.na(universe) & universe != ""]

# Get DESeq files
deseq_files <- list.files(
  path = file.path(deseq_outDir, "sigDEgeneLists"),
  pattern = "^filteredDEG.*_noNA\\.tab$",  # starts with 'filteredDEG' and ends with '_noNA.tab'
  full.names = TRUE
)

# Clean gene names in input lists
setbp1_targets <- clean_gene_names(setbp1_targets)
ID_genes <- clean_gene_names(ID_genes)
autism_genes <- clean_gene_names(autism_genes)
ectopicG870S <- clean_gene_names(ectopicG870S) #now there are no more tabs in front of gene names
universe <- clean_gene_names(universe)

## only expressed ones
setbp1_targets <- intersect(setbp1_targets, universe)
ID_genes <- intersect(ID_genes, universe)
autism_genes <- intersect(autism_genes, universe)
ectopicG870S <- intersect(ectopicG870S, universe)
total_panelapp <- unlist(ID_genes, autism_genes)
total_panelapp <- unique(total_panelapp)

# Define the universe size (this should be the total number of genes considered in your DESeq analysis)
universe_size <- nrow(res1_deseq_results_noNA)  

# Check overlap for ID & autism genes
overlap_ID_autism <- intersect(ID_genes, autism_genes)
if (length(overlap_ID_autism) > 0) {
  write.csv(overlap_ID_autism, 
            file = paste0(deseq_outDir, "overlaps/2-set/ID_autism_overlap.csv"),
            quote = F, row.names = F)
  
  plot_venn_pdf(
    list(ID_genes, autism_genes),
    c("ID Genes", "Autism Genes"),
    paste0(deseq_outDir, "overlaps/2-set/ID_autism_venn.pdf")
  )
  
  p_value <- perform_fishers_test(ID_genes, autism_genes, universe_size)
  write.csv(data.frame(p_value = p_value), 
            file = paste0(deseq_outDir, "overlaps/2-set/ID_autism_fishers_pvalue.csv"),
            quote = F, row.names = F)
} else {
  print(paste("No overlap for ID & autism genes"))
}

# Process DESeq files
for (i in deseq_files) {
  if (i %in% 4:6) next  # Skip files 4, 5, 6
  tmp_f <- read.delim(i, header = TRUE)  # still read with header=TRUE, but it's just one column
  
  tmp_f_name <- sub("\\.tab$", "", basename(i))

  # Extract and clean symbol column as a vector
  DEGs <- tmp_f$symbol
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]           # Remove NA and empty
  DEGs <- clean_gene_names(DEGs)                      # Apply cleaning function
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]           # Remove NA again if cleaning added any
  

  # Check overlap for setbp1 targets
  overlap_setbp1 <- intersect(setbp1_targets, DEGs)
  if (length(overlap_setbp1) > 0) {
    write.csv(overlap_setbp1, 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_overlap_", tmp_f_name,".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(setbp1_targets, DEGs),
      c("SETBP1 Targets", tmp_f_name),
      paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_venn_", tmp_f_name, ".pdf")
    )
    
    p_value <- perform_fishers_test(setbp1_targets, DEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_fishers_pvalue_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for setbp1 targets in file", tmp_f_name))
  }
  
  # Check overlap for ID genes
  overlap_ID_genes <- intersect(ID_genes, DEGs)
  if (length(overlap_ID_genes) > 0) {
    write.csv(overlap_ID_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_IDgenes_overlap_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ID_genes, DEGs),
      c("ID Genes", tmp_f_name),
      paste0(deseq_outDir, "overlaps/2-set/ID_genes_venn_", tmp_f_name, ".pdf")
    )
    
    p_value <- perform_fishers_test(ID_genes, DEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ID_genes_fishers_pvalue_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ID genes in file", tmp_f_name))
  }
  
    # Check overlap for autism genes
  overlap_autism_genes <- intersect(autism_genes, DEGs)
  if (length(overlap_autism_genes) > 0) {
    write.csv(overlap_autism_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_autism_genes_overlap_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(autism_genes, DEGs),
      c("Autism Genes", tmp_f_name),
      paste0(deseq_outDir, "overlaps/2-set/autism_genes_venn_", tmp_f_name, ".pdf")
    )
    
    p_value <- perform_fishers_test(autism_genes, DEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/autism_genes_fishers_pvalue_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for autism genes in file", tmp_f_name))
  }
  
    # Check overlap for all panelapp genes
  overlap_total_panelapp <- intersect(total_panelapp, DEGs)
  if (length(overlap_total_panelapp) > 0) {
    write.csv(overlap_total_panelapp, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_total_panelapp_overlap_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(total_panelapp, DEGs),
      c("Panelapp ID & Autism Genes", tmp_f_name),
      paste0(deseq_outDir, "overlaps/2-set/total_panelapp_venn_", tmp_f_name, ".pdf")
    )
    
    p_value <- perform_fishers_test(total_panelapp, DEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/total_panelapp_fishers_pvalue_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for total panelapp genes in file", tmp_f_name))
  }
  
  # Check overlap for ectopicG870S
  overlap_ectopicG870S <- intersect(ectopicG870S, DEGs)
  if (length(overlap_ectopicG870S) > 0) {
    write.csv(overlap_ectopicG870S, 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_DEG_overlap_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ectopicG870S, DEGs),
      c("Ectopic G870S", tmp_f_name),
      paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_venn_", tmp_f_name, ".pdf")
    )
    
    p_value <- perform_fishers_test(ectopicG870S, DEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_fishers_pvalue_", tmp_f_name, ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ectopicG870S in file", tmp_f_name))
  }
  
  summary(DEGs)
}
```
# 2-set venn
## overlap of sig DEGs
```{r}
# Check overlap for sig DEG in all conditions with the lists
# ## Get the intersection of all lists
gene_lists <- list()

for (i in deseq_files) {
  if (i %in% 4:6) next  # skip files 4, 5, 6
  tmp_f <- read.delim(i, header = TRUE)
  tmp_f_name <- sub("\\.tab$", "", basename(i))
  
  DEGs <- tmp_f$symbol
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  DEGs <- clean_gene_names(DEGs)
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  
  gene_lists[[tmp_f_name]] <- DEGs
}

# Now get the shared genes across all DEG lists
sigDEG_allSETBP1 <- Reduce(intersect, gene_lists)

cat("Number of shared DEGs across all included files:", length(sigDEG_allSETBP1), "\n")
print(sigDEG_allSETBP1)

# Save overlap as CSV
write.csv(
  data.frame(Genes = sigDEG_allSETBP1),
  file = file.path(deseq_outDir, "overlap_allSignDEGs_res1-3.csv"),
  quote = FALSE, row.names = FALSE)


  # Check overlap for setbp1 targets
  overlap_setbp1 <- intersect(setbp1_targets, sigDEG_allSETBP1)
  if (length(overlap_setbp1) > 0) {
    write.csv(overlap_setbp1, 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_overlap_sigDEG_allSETBP1.csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(setbp1_targets, sigDEG_allSETBP1),
      c("SETBP1 Targets", "sigDEG_allSETBP1"),
      paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_venn_sigDEG_allSETBP1", ".pdf")
    )
    
    p_value <- perform_fishers_test(setbp1_targets, sigDEG_allSETBP1, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_fishers_pvalue_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for setbp1 targets in file sigDEG_allSETBP1"))
  }
  
  # Check overlap for ID genes
  overlap_ID_genes <- intersect(ID_genes, sigDEG_allSETBP1)
  if (length(overlap_ID_genes) > 0) {
    write.csv(overlap_ID_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_IDgenes_overlap_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ID_genes, sigDEG_allSETBP1),
      c("ID Genes", "sigDEG_allSETBP1"),
      paste0(deseq_outDir, "overlaps/2-set/ID_genes_venn_sigDEG_allSETBP1", ".pdf")
    )
    
    p_value <- perform_fishers_test(ID_genes, sigDEG_allSETBP1, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ID_genes_fishers_pvalue_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ID genes in file sigDEG_allSETBP1"))
  }
  
  # Check overlap for autism genes
  overlap_autism_genes <- intersect(autism_genes, sigDEG_allSETBP1)
  if (length(overlap_autism_genes) > 0) {
    write.csv(overlap_autism_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_autism_genes_overlap_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(autism_genes, sigDEG_allSETBP1),
      c("Autism Genes", "sigDEG_allSETBP1"),
      paste0(deseq_outDir, "overlaps/2-set/autism_genes_venn_sigDEG_allSETBP1", ".pdf")
    )
    
    p_value <- perform_fishers_test(autism_genes, sigDEG_allSETBP1, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/autism_genes_fishers_pvalue_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for autism genes in file sigDEG_allSETBP1"))
  }
  
  # Check overlap for total panelapp genes
  overlap_total_panelapp <- intersect(total_panelapp, sigDEG_allSETBP1)
  if (length(overlap_total_panelapp) > 0) {
    write.csv(overlap_total_panelapp, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_total_panelapp_overlap_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(total_panelapp, sigDEG_allSETBP1),
      c("Panelapp ID & Autism Genes", "sigDEG_allSETBP1"),
      paste0(deseq_outDir, "overlaps/2-set/total_panelapp_venn_sigDEG_allSETBP1", ".pdf")
    )
    
    p_value <- perform_fishers_test(total_panelapp, sigDEG_allSETBP1, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/total_panelapp_fishers_pvalue_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for autism genes in file sigDEG_allSETBP1"))
  }
  
  overlap_ectopicG870S <- intersect(ectopicG870S, sigDEG_allSETBP1)
  if (length(overlap_ectopicG870S) > 0) {
    write.csv(overlap_ectopicG870S, 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_DEG_overlap_sigDEG_allSETBP1",".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ectopicG870S, sigDEG_allSETBP1),
      c("Ectopic G870S", "sigDEG_allSETBP1"),
      paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_venn_sigDEG_allSETBP1", ".pdf")
    )
    
    p_value <- perform_fishers_test(ectopicG870S, sigDEG_allSETBP1, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_fishers_pvalue_sigDEG_allSETBP1", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ectopicG870S in file sigDEG_allSETBP1"))
  }
```

## total sig DEGs
```{r}
# Check total sig DEG in all conditions with the lists

# Initialize an empty vector to collect all genes

gene_lists <- list()

for (i in deseq_files) {
  if (i %in% 4:6) next  # skip files 4, 5, 6
  tmp_f <- read.delim(i, header = TRUE)
  tmp_f_name <- sub("\\.tab$", "", basename(i))
  
  DEGs <- tmp_f$symbol
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  DEGs <- clean_gene_names(DEGs)
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  
  gene_lists[[tmp_f_name]] <- DEGs
}

# Now get the shared genes across all DEG lists
total_unique_sigDEGs <- unique(unlist(gene_lists))

cat("Number of shared DEGs across all included files:", length(total_unique_sigDEGs), "\n")
print(total_unique_sigDEGs)

# Save overlap as CSV
write.csv(
  data.frame(Genes = total_unique_sigDEGs),
  file = file.path(deseq_outDir, "total_unique_SignDEGs_res1-3.csv"),
  quote = FALSE, row.names = FALSE)


  # Check overlap for setbp1 targets
  overlap_setbp1 <- intersect(setbp1_targets, total_unique_sigDEGs)
  if (length(overlap_setbp1) > 0) {
    write.csv(overlap_setbp1, 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_overlap_sigDEG_allSETBP1",".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(setbp1_targets, total_unique_sigDEGs),
      c("SETBP1 Targets", "total_unique_sigDEGs"),
      paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_venn_total_unique_sigDEGs", ".pdf")
    )
    
    p_value <- perform_fishers_test(setbp1_targets, total_unique_sigDEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/setbp1Targets_fishers_pvaluetotal_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for setbp1 targets in file total_unique_sigDEGs"))
  }
  
  # Check overlap for ID genes
  overlap_ID_genes <- intersect(ID_genes, total_unique_sigDEGs)
  if (length(overlap_ID_genes) > 0) {
    write.csv(overlap_ID_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_IDgenes_overlaptotal_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ID_genes, total_unique_sigDEGs),
      c("ID Genes", "total_unique_sigDEGs"),
      paste0(deseq_outDir, "overlaps/2-set/ID_genes_venn_total_unique_sigDEGs", ".pdf")
    )
    
    p_value <- perform_fishers_test(ID_genes, total_unique_sigDEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ID_genes_fishers_pvaluetotal_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ID genes in file total_unique_sigDEGs"))
  }
  
  # Check overlap for autism genes
  overlap_autism_genes <- intersect(autism_genes, total_unique_sigDEGs)
  if (length(overlap_autism_genes) > 0) {
    write.csv(overlap_autism_genes, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_autism_genes_overlap_total_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(autism_genes, total_unique_sigDEGs),
      c("Autism Genes", "total_unique_sigDEGs"),
      paste0(deseq_outDir, "overlaps/2-set/autism_genes_venn_total_unique_sigDEGs", ".pdf")
    )
    
    p_value <- perform_fishers_test(autism_genes, total_unique_sigDEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/autism_genes_fishers_pvalue_total_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for autism genes in file total_unique_sigDEGs"))
  }
  
  # Check overlap for total panelapp genes
  overlap_total_panelapp <- intersect(total_panelapp, total_unique_sigDEGs)
  if (length(overlap_total_panelapp) > 0) {
    write.csv(overlap_total_panelapp, 
              file = paste0(deseq_outDir, "overlaps/2-set/panelapp_total_panelapp_overlap_total_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(total_panelapp, total_unique_sigDEGs),
      c("Panelapp ID & Autism Genes", "total_unique_sigDEGs"),
      paste0(deseq_outDir, "overlaps/2-set/total_panelapp_venn_total_unique_sigDEGs", ".pdf")
    )
    
    p_value <- perform_fishers_test(total_panelapp, total_unique_sigDEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/total_panelapp_fishers_pvalue_total_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for autism genes in file total_unique_sigDEGs"))
  }
  
  overlap_ectopicG870S <- intersect(ectopicG870S, total_unique_sigDEGs)
  if (length(overlap_ectopicG870S) > 0) {
    write.csv(overlap_ectopicG870S, 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_DEG_overlap_total_unique_sigDEGs",".csv"),
              quote = F, row.names = F)
    
    plot_venn_pdf(
      list(ectopicG870S, total_unique_sigDEGs),
      c("Ectopic G870S", "total_unique_sigDEGs"),
      paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_venn_total_unique_sigDEGs", ".pdf")
    )
    
    p_value <- perform_fishers_test(ectopicG870S, total_unique_sigDEGs, universe_size)
    write.csv(data.frame(p_value = p_value), 
              file = paste0(deseq_outDir, "overlaps/2-set/ectopicG870S_fishers_pvalue_total_unique_sigDEGs", ".csv"),
              quote = F, row.names = F)
  } else {
    print(paste("No overlap for ectopicG870S in file total_unique_sigDEGs"))
  }
```

# for 3-set venn
## overlap of sigDEGs and total sigDEGs with target gene lists
```{r}
# function to plot 3-set venn
plot_venn_pdf <- function(gene_lists, set_names, output_file, fill = c("orange", "green3", "skyblue2")) {
  #library(VennDiagram)

  A <- gene_lists[[1]]
  B <- gene_lists[[2]]
  C <- gene_lists[[3]]

  venn <- VennDiagram::draw.triple.venn(
    area1 = length(A),
    area2 = length(B),
    area3 = length(C),
    n12 = length(intersect(A, B)),
    n23 = length(intersect(B, C)),
    n13 = length(intersect(A, C)),
    n123 = length(Reduce(intersect, list(A, B, C))),
    category = set_names,
    fill = fill,
    lty = "blank",
    cex = 1.5,
    cat.cex = 1.5,
    cat.pos = c(-20, 20, 0),
    cat.dist = 0.07
  )

  pdf(output_file, width = 7, height = 7)
  grid::grid.draw(venn)
  dev.off()
}

# to find overlap of 3 gene lists and plot venn diagramme  
# Combine lists
# first for overlap of sig DEGs
gene_lists <- list(
  ID_genes = ID_genes,
  Autism_genes = autism_genes,
  sigDEG_allSETBP1 = sigDEG_allSETBP1
)

# now plot
plot_venn_pdf(
  gene_lists,
  c("ID Genes", "Autism Genes", "sigDEG_allSETBP1"),
  paste0(deseq_outDir, "overlaps/3-set/ID_Autism_overlapSigDEG_venn.pdf"),
  fill = c("orange", "green3", "skyblue2")  # exactly 3 colors for 3 sets
)

# Compute overlaps
overlap_ID_autism <- intersect(ID_genes, autism_genes)
overlap_ID_sigDEG <- intersect(ID_genes, sigDEG_allSETBP1)
overlap_autism_sigDEG <- intersect(autism_genes, sigDEG_allSETBP1)
overlap_all_three <- Reduce(intersect, gene_lists)

# Write overlaps to files
write.csv(overlap_ID_autism,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_ID_sigDEG,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_overlapSigDEG_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_autism_sigDEG,
          file = paste0(deseq_outDir, "overlaps/3-set/autism_overlapSigDEG_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_all_three,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_overlapSigDEG_all_three_overlap.csv"),
          quote = FALSE, row.names = FALSE)

# Perform Fisher's exact tests (pairwise only — not for 3-way)
pval_ID_sigDEG <- perform_fishers_test(ID_genes, sigDEG_allSETBP1, universe_size)
pval_autism_sigDEG <- perform_fishers_test(autism_genes, sigDEG_allSETBP1, universe_size)
pval_ID_autism <- perform_fishers_test(ID_genes, autism_genes, universe_size)

# Save p-values
write.csv(data.frame(p_value = pval_ID_sigDEG),
          file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_overlapSigDEG_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(data.frame(p_value = pval_autism_sigDEG),
          file = paste0(deseq_outDir, "overlaps/3-set/autism_vs_overlapSigDEG_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(data.frame(p_value = pval_ID_autism),
          file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_autism_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)


# to find overlap of 3 gene lists and plot venn diagramme 
# Combine lists
# total sig DEGs
gene_lists <- list(
  ID_genes = ID_genes,
  Autism_genes = autism_genes,
  sigDEG_allSETBP1 = total_unique_sigDEGs
)

# now plot
plot_venn_pdf(
  gene_lists,
  c("ID Genes", "Autism Genes", "total unique sigDEGs"),
  paste0(deseq_outDir, "overlaps/3-set/ID_Autism_sigDEG_total_unique_sigDEGs_venn.pdf"),
  fill = c("orange", "green3", "skyblue2")  # exactly 3 colors for 3 sets
)

# Compute overlaps
overlap_ID_autism <- intersect(ID_genes, autism_genes)
overlap_ID_sigDEG <- intersect(ID_genes, total_unique_sigDEGs)
overlap_autism_sigDEG <- intersect(autism_genes, total_unique_sigDEGs)
overlap_all_three <- Reduce(intersect, gene_lists)

# Write overlaps to files
write.csv(overlap_ID_autism,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_ID_sigDEG,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_totalSigDEG_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_autism_sigDEG,
          file = paste0(deseq_outDir, "overlaps/3-set/autism_totalSigDEG_overlap.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(overlap_all_three,
          file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_totalSIgDEG_all_three_overlap.csv"),
          quote = FALSE, row.names = FALSE)

# Perform Fisher's exact tests (pairwise only — not for 3-way)
pval_ID_sigDEG <- perform_fishers_test(ID_genes, total_unique_sigDEGs, universe_size)
pval_autism_sigDEG <- perform_fishers_test(autism_genes, total_unique_sigDEGs, universe_size)
pval_ID_autism <- perform_fishers_test(ID_genes, autism_genes, universe_size)

# Save p-values
write.csv(data.frame(p_value = pval_ID_sigDEG),
          file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_totalSigDEG_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(data.frame(p_value = pval_autism_sigDEG),
          file = paste0(deseq_outDir, "overlaps/3-set/autism_vs_totalSigDEG_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)

write.csv(data.frame(p_value = pval_ID_autism),
          file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_autism_fishers_pvalue.csv"),
          quote = FALSE, row.names = FALSE)
```

## 3-set venn for each res
```{r}

# for each res
for (i in deseq_files) {
  if (i %in% 4:6) next  # skip files 4, 5, 6
  tmp_f <- read.delim(i, header = TRUE)
  tmp_f_name <- sub("\\.tab$", "", basename(i))
  
  DEGs <- tmp_f$symbol
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  DEGs <- clean_gene_names(DEGs)
  DEGs <- DEGs[!is.na(DEGs) & DEGs != ""]
  
  gene_lists <- list(
    ID_genes = ID_genes,
    Autism_genes = autism_genes,
    DEG = DEGs
  )
  
  plot_venn_pdf(
    gene_lists,
    c("ID Genes", "Autism Genes", tmp_f_name),
    paste0(deseq_outDir, "overlaps/3-set/ID_Autism_DEG_venn_", tmp_f_name, ".pdf")
  )
  
  overlap_ID_autism <- intersect(ID_genes, autism_genes)
  overlap_ID_DEG <- intersect(ID_genes, DEGs)
  overlap_autism_DEG <- intersect(autism_genes, DEGs)
  overlap_all_three <- Reduce(intersect, gene_lists)
  
  write.csv(overlap_ID_autism,
            file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_overlap_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  write.csv(overlap_ID_DEG,
            file = paste0(deseq_outDir, "overlaps/3-set/ID_DEG_overlap_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  write.csv(overlap_autism_DEG,
            file = paste0(deseq_outDir, "overlaps/3-set/autism_DEG_overlap_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  write.csv(overlap_all_three,
            file = paste0(deseq_outDir, "overlaps/3-set/ID_autism_DEG_all_three_overlap_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  pval_ID_DEG <- perform_fishers_test(ID_genes, DEGs, universe_size)
  pval_autism_DEG <- perform_fishers_test(autism_genes, DEGs, universe_size)
  pval_ID_autism <- perform_fishers_test(ID_genes, autism_genes, universe_size)
  
  write.csv(data.frame(p_value = pval_ID_DEG),
            file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_DEG_fishers_pvalue_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  write.csv(data.frame(p_value = pval_autism_DEG),
            file = paste0(deseq_outDir, "overlaps/3-set/autism_vs_DEG_fishers_pvalue_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  write.csv(data.frame(p_value = pval_ID_autism),
            file = paste0(deseq_outDir, "overlaps/3-set/ID_vs_autism_fishers_pvalue_", tmp_f_name, ".csv"),
            quote = FALSE, row.names = FALSE)
  
  summary(DEGs)
}




```


```{r session info}
save.image(file = paste0(dir, "lg-scrna/working_data/Maggie/working/script/bulk_rna_SETBP1_HDF/from_gokberk/20250528/setbp1_hdf_rna_20250528_final.RData"))
sessionInfo()

```

